Create a setting for user to indicate humanized or hours on dashboard/device detail

bauzer714 opened the pull request:
Default is "humanize" creates a non-breaking change.

**Setting:**
![image](https://github.com/AnalogJ/scrutiny/assets/13224936/cd34cdd1-fa4d-4ade-96dc-fbbf2ac6b484)

![image](https://github.com/AnalogJ/scrutiny/assets/13224936/45411419-eed1-47ed-859d-0ed5d53a0077)

Dashboard:
![image](https://github.com/AnalogJ/scrutiny/assets/13224936/5df2c7bb-5202-45b4-bd22-415fa0b1a8f3)

![image](https://github.com/AnalogJ/scrutiny/assets/13224936/9ea9e0a1-77ae-4552-bf77-1ee440f28900)

Detail:
![image](https://github.com/AnalogJ/scrutiny/assets/13224936/903ea281-dd47-4757-b11a-5471cf98a2b0)

Detail tooltip:
![image](https://github.com/AnalogJ/scrutiny/assets/13224936/c82f6677-0e6d-488c-a9ed-4137e05ffead)

![image](https://github.com/AnalogJ/scrutiny/assets/13224936/e9a237da-a0c5-4405-9f65-5e5a6b024b19)


Implements original request of #628.

Commit:
Add DB Migration and config/settings

Commit:
Add pipe and implement to dashboard/device component

Commit:
Update db migration description to match setting possible values

AnalogJ commented on webapp/backend/pkg/database/scrutiny_repository_migrations.go:
```diff
@@ -332,6 +332,12 @@ func (sr *scrutinyRepository) Migrate(ctx context.Context) error {
+						SettingKeyDescription: "Presentation format for device powered on time ('humanize' | 'device_hours')",
+						SettingDataType:       "string",
+						SettingValueString:    "humanize",
+					},
```
This has to be added as a *new* migration, otherwise it will be skipped for users with an existing database..

AnalogJ commented:
sure, I can get behind this. 
thanks for the PR! 🥳 

Added comment about the db settings change - it needs to be in it's own migration

bauzer714 commented:
Thanks @AnalogJ - I've never developed in Go, any specific naming conventions or versioning? Does it need to be removed from the current script?

AnalogJ commented:
Basically just copy paste a migration entry like this: https://github.com/bauzer714/scrutiny/blob/addDeviceHoursSetting/webapp/backend/pkg/database/scrutiny_repository_migrations.go#L379C3-L393C5

The ID should start with "m" and be followed by a timestamp in YYYYMMDDHHMMSS format

`date '+%Y%m%d%H%M%S'` should do it

Thanks @bauzer714 !

Commit:
Move powered_on_hours_unit to a new migration id

bauzer714 commented:
@AnalogJ - thanks for the guidance. Pushed the latest changes. Let me know if this needs further changes.

AnalogJ commented on webapp/frontend/src/app/shared/device-hours.pipe.spec.ts:
```diff
@@ -0,0 +1,9 @@
+    const pipe = new DeviceHoursPipe();
+    expect(pipe).toBeTruthy();
+  });
+});
```
one last nit. Can you add some tests to show how this pipe behaves when called with:

- 'humanize'
- 'device_hours'
- null
- a invalid string

thanks! @bauzer714  this is great work

Commit:
Add tests for DeviceHoursPipe

bauzer714 commented:
@AnalogJ  - updated.  I added tests for "configuration" of each of the 4 requested.

Additionally I added null input values (since null is a number in TS) for the configuration "device_hours" and "humanize". 

![image](https://github.com/user-attachments/assets/87c5a732-0dcb-499a-9efd-bc9cec720e6d)

What's your thoughts on "null hours" vs "0 hours" as the expected output of the "null" input tests? A few options I thought of:

- Leave both as is
- Update the device_hours pipe to return '0 hours'
- Update all configurations of the pipe to return ''
- ???

![image](https://github.com/user-attachments/assets/601d240e-904c-4314-8518-5af370758368)

AnalogJ commented:
if no data is present, maybe "No Data" or "Unknown"

Commit:
Adjust null input response, and tests

bauzer714 commented:
@AnalogJ  - Updated.   Let me know any additional changes.

![image](https://github.com/user-attachments/assets/917f6397-5cca-4949-bf61-c5d46f7e79ca)

AnalogJ commented:
LGTM. Thanks for the PR @bauzer714 ! 🥇

AnalogJ committed:
Merge branch 'master' into addDeviceHoursSetting
